File: internal/model/condition.go
package model

import (
	"fmt"
	"strings"

	"gopkg.in/yaml.v3"
)

// Operand kinds: a condition tree is built from operands that are either
// path references into the payload, literal values, or nested operations.
type OperandKind int

const (
	OperandPath    OperandKind = iota // bare identifier — resolved against payload
	OperandLiteral                    // quoted string, number, boolean — used as-is
	OperandOp                         // nested operation
)

// Operand is one argument to an operator. Exactly one of Path/Literal/Op is set.
type Operand struct {
	Kind    OperandKind
	Path    string  // dot-separated payload path, e.g. "nested.count"
	Literal any     // string, int64, float64, bool, nil
	Op      *OpExpr // nested operation
}

// OpExpr is a single operator with its operands, e.g. eq(count, "3").
type OpExpr struct {
	Op       string    // "eq", "neq", "gt", "gte", "lt", "lte", "len",
	                    // "and", "or", "not", "exists"
	Operands []Operand
}

// Condition is the top-level condition tree on a Filter branch.
// Per the design: a top-level list of OpExprs is implicit OR.
// A single OpExpr is just that condition.
type Condition struct {
	Exprs []OpExpr // implicit OR if len > 1
}

// IsEmpty reports whether the condition is unconditional (always true).
func (c *Condition) IsEmpty() bool {
	return c == nil || len(c.Exprs) == 0
}

// Pretty returns a human-readable form for error messages.
func (c *Condition) Pretty() string {
	if c.IsEmpty() {
		return "true"
	}
	parts := make([]string, 0, len(c.Exprs))
	for _, e := range c.Exprs {
		parts = append(parts, prettyOp(&e))
	}
	if len(parts) == 1 {
		return parts[0]
	}
	return "(" + strings.Join(parts, " or ") + ")"
}

func prettyOp(op *OpExpr) string {
	if op == nil {
		return "<nil>"
	}
	args := make([]string, 0, len(op.Operands))
	for _, o := range op.Operands {
		args = append(args, prettyOperand(&o))
	}
	return fmt.Sprintf("%s(%s)", op.Op, strings.Join(args, ", "))
}

func prettyOperand(o *Operand) string {
	switch o.Kind {
	case OperandPath:
		return o.Path
	case OperandLiteral:
		switch v := o.Literal.(type) {
		case string:
			return fmt.Sprintf("%q", v)
		default:
			return fmt.Sprintf("%v", v)
		}
	case OperandOp:
		return prettyOp(o.Op)
	}
	return "<?>"
}

// --- YAML unmarshaling ----------------------------------------------------
//
// YAML form:
//   condition:
//     - eq: [count, "3"]              # single op, args inline
//     - eq:                            # single op, args block
//         - count
//         - "3"
//     - eq:                            # nested op
//         - len: [someList]
//         - "5"
//     - and:
//         - eq: [a, "x"]
//         - eq: [b, "y"]
//
// Scalar style determines path vs literal:
//   - Plain (unquoted) string scalars  → OperandPath
//   - Single/double-quoted strings     → OperandLiteral (string)
//   - Numeric scalars (no quotes)      → OperandLiteral (int64/float64)
//   - Boolean scalars                  → OperandLiteral (bool)

// UnmarshalYAML implements custom parsing for Condition.
func (c *Condition) UnmarshalYAML(node *yaml.Node) error {
	if node == nil {
		c.Exprs = nil
		return nil
	}
	if node.Kind != yaml.SequenceNode {
		return fmt.Errorf("condition must be a YAML list, got %s", kindName(node.Kind))
	}
	c.Exprs = make([]OpExpr, 0, len(node.Content))
	for i, item := range node.Content {
		expr, err := parseOpExpr(item)
		if err != nil {
			return fmt.Errorf("condition[%d]: %w", i, err)
		}
		c.Exprs = append(c.Exprs, *expr)
	}
	return nil
}

func parseOpExpr(node *yaml.Node) (*OpExpr, error) {
	if node.Kind != yaml.MappingNode {
		return nil, fmt.Errorf("expected operator mapping, got %s", kindName(node.Kind))
	}
	if len(node.Content) != 2 {
		return nil, fmt.Errorf("operator must have exactly one key (got %d entries)", len(node.Content)/2)
	}
	keyNode := node.Content[0]
	valNode := node.Content[1]
	if keyNode.Kind != yaml.ScalarNode {
		return nil, fmt.Errorf("operator name must be a scalar")
	}
	op := &OpExpr{Op: keyNode.Value}
	operands, err := parseOperandList(valNode)
	if err != nil {
		return nil, fmt.Errorf("op %q: %w", op.Op, err)
	}
	op.Operands = operands
	return op, nil
}

// parseOperandList accepts a YAML sequence node where each item is an operand
// (path, literal, or nested op). It also accepts a single scalar/mapping in
// place of a one-item sequence for tolerance.
func parseOperandList(node *yaml.Node) ([]Operand, error) {
	if node.Kind == yaml.SequenceNode {
		out := make([]Operand, 0, len(node.Content))
		for i, item := range node.Content {
			operand, err := parseOperand(item)
			if err != nil {
				return nil, fmt.Errorf("operand[%d]: %w", i, err)
			}
			out = append(out, *operand)
		}
		return out, nil
	}
	// single operand allowed for unary ops (e.g. not, len, exists).
	one, err := parseOperand(node)
	if err != nil {
		return nil, err
	}
	return []Operand{*one}, nil
}

func parseOperand(node *yaml.Node) (*Operand, error) {
	switch node.Kind {
	case yaml.ScalarNode:
		return parseScalarOperand(node)
	case yaml.MappingNode:
		// nested operator
		expr, err := parseOpExpr(node)
		if err != nil {
			return nil, err
		}
		return &Operand{Kind: OperandOp, Op: expr}, nil
	default:
		return nil, fmt.Errorf("unsupported operand kind: %s", kindName(node.Kind))
	}
}

func parseScalarOperand(node *yaml.Node) (*Operand, error) {
	// YAML scalar style determines path vs literal.
	// Plain  (unquoted) string → path
	// Quoted string → literal string
	// Tagged numeric/bool → literal of that type
	switch node.Style {
	case yaml.SingleQuotedStyle, yaml.DoubleQuotedStyle:
		return &Operand{Kind: OperandLiteral, Literal: node.Value}, nil
	}

	// Plain scalar: defer to YAML's resolved tag.
	switch node.Tag {
	case "!!int":
		var v int64
		if err := node.Decode(&v); err != nil {
			return nil, fmt.Errorf("invalid integer literal: %w", err)
		}
		return &Operand{Kind: OperandLiteral, Literal: v}, nil
	case "!!float":
		var v float64
		if err := node.Decode(&v); err != nil {
			return nil, fmt.Errorf("invalid float literal: %w", err)
		}
		return &Operand{Kind: OperandLiteral, Literal: v}, nil
	case "!!bool":
		var v bool
		if err := node.Decode(&v); err != nil {
			return nil, fmt.Errorf("invalid bool literal: %w", err)
		}
		return &Operand{Kind: OperandLiteral, Literal: v}, nil
	case "!!null":
		return &Operand{Kind: OperandLiteral, Literal: nil}, nil
	}

	// Plain string → path reference into payload.
	if node.Value == "" {
		return nil, fmt.Errorf("empty path operand")
	}
	return &Operand{Kind: OperandPath, Path: node.Value}, nil
}

func kindName(k yaml.Kind) string {
	switch k {
	case yaml.DocumentNode:
		return "document"
	case yaml.SequenceNode:
		return "sequence"
	case yaml.MappingNode:
		return "mapping"
	case yaml.ScalarNode:
		return "scalar"
	case yaml.AliasNode:
		return "alias"
	}
	return fmt.Sprintf("unknown(%d)", k)
}

File: sdk/typescript/src/condition.ts
/**
 * Condition evaluator. Mirrors the Go reference implementation in
 * internal/model/eval.go.
 *
 * Top-level Condition.exprs are combined with implicit OR.
 * Operators: eq neq gt gte lt lte len and or not exists
 * Loose-typed equality coerces numeric strings to numbers ("3" == 3 ⇒ true).
 */

import type {
  Condition,
  OpExpr,
  Operand,
  Payload,
} from "./types.js";

export function evaluate(c: Condition | undefined, payload: Payload): boolean {
  if (!c || !c.exprs || c.exprs.length === 0) return true;
  for (const expr of c.exprs) {
    if (truthy(evalOp(expr, payload))) return true;
  }
  return false;
}

function evalOperand(o: Operand, payload: Payload): unknown {
  switch (o.kind) {
    case "literal":
      return o.value;
    case "path":
      return resolvePath(payload, o.path);
    case "op":
      return evalOp(o.op, payload);
  }
}

function resolvePath(payload: Payload, path: string): unknown {
  if (!path) return undefined;
  const parts = path.split(".");
  let cur: unknown = payload;
  for (const p of parts) {
    if (cur === null || cur === undefined) return undefined;
    if (typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[p];
  }
  return cur;
}

function evalOp(op: OpExpr, payload: Payload): unknown {
  switch (op.op) {
    case "eq":
      return binCmp(op, payload, looseEqual);
    case "neq":
      return binCmp(op, payload, (a, b) => !looseEqual(a, b));
    case "gt":
      return numCmp(op, payload, (a, b) => a > b);
    case "gte":
      return numCmp(op, payload, (a, b) => a >= b);
    case "lt":
      return numCmp(op, payload, (a, b) => a < b);
    case "lte":
      return numCmp(op, payload, (a, b) => a <= b);
    case "len":
      return evalLen(op, payload);
    case "and":
      return evalAndOr(op, payload, true);
    case "or":
      return evalAndOr(op, payload, false);
    case "not":
      return evalNot(op, payload);
    case "exists":
      return evalExists(op, payload);
    default:
      throw new Error(`unknown operator "${op.op}"`);
  }
}

function binCmp(
  op: OpExpr,
  payload: Payload,
  fn: (a: unknown, b: unknown) => boolean,
): boolean {
  if (op.operands.length !== 2) {
    throw new Error(`${op.op} requires 2 operands, got ${op.operands.length}`);
  }
  return fn(evalOperand(op.operands[0]!, payload), evalOperand(op.operands[1]!, payload));
}

function numCmp(
  op: OpExpr,
  payload: Payload,
  fn: (a: number, b: number) => boolean,
): boolean {
  if (op.operands.length !== 2) {
    throw new Error(`${op.op} requires 2 operands, got ${op.operands.length}`);
  }
  const a = evalOperand(op.operands[0]!, payload);
  const b = evalOperand(op.operands[1]!, payload);
  const af = toNumber(a);
  const bf = toNumber(b);
  if (af === undefined) {
    throw new Error(`${op.op}: left operand ${stringify(a)} is not numeric`);
  }
  if (bf === undefined) {
    throw new Error(`${op.op}: right operand ${stringify(b)} is not numeric`);
  }
  return fn(af, bf);
}

function evalLen(op: OpExpr, payload: Payload): number {
  if (op.operands.length !== 1) {
    throw new Error(`len requires 1 operand, got ${op.operands.length}`);
  }
  const v = evalOperand(op.operands[0]!, payload);
  if (typeof v === "string") return v.length;
  if (Array.isArray(v)) return v.length;
  if (v === null || v === undefined) return 0;
  if (typeof v === "object") return Object.keys(v as object).length;
  throw new Error(`len: unsupported type ${typeof v}`);
}

function evalAndOr(op: OpExpr, payload: Payload, isAnd: boolean): boolean {
  if (op.operands.length === 0) return isAnd; // empty AND = true; empty OR = false
  for (const o of op.operands) {
    const t = truthy(evalOperand(o, payload));
    if (isAnd && !t) return false;
    if (!isAnd && t) return true;
  }
  return isAnd;
}

function evalNot(op: OpExpr, payload: Payload): boolean {
  if (op.operands.length !== 1) {
    throw new Error(`not requires 1 operand, got ${op.operands.length}`);
  }
  return !truthy(evalOperand(op.operands[0]!, payload));
}

function evalExists(op: OpExpr, payload: Payload): boolean {
  if (op.operands.length !== 1) {
    throw new Error(`exists requires 1 operand, got ${op.operands.length}`);
  }
  const o = op.operands[0]!;
  if (o.kind !== "path") {
    throw new Error("exists: operand must be a path");
  }
  const v = resolvePath(payload, o.path);
  return v !== undefined && v !== null;
}

// --- helpers --------------------------------------------------------------

function truthy(v: unknown): boolean {
  if (v === null || v === undefined) return false;
  if (typeof v === "boolean") return v;
  if (typeof v === "number") return v !== 0;
  if (typeof v === "string") return v.length > 0;
  if (Array.isArray(v)) return v.length > 0;
  if (typeof v === "object") return Object.keys(v as object).length > 0;
  return true;
}

function looseEqual(a: unknown, b: unknown): boolean {
  if (a === null || a === undefined || b === null || b === undefined) {
    return (a === null || a === undefined) && (b === null || b === undefined);
  }
  const af = toNumber(a);
  const bf = toNumber(b);
  if (af !== undefined && bf !== undefined) return af === bf;
  return stringify(a) === stringify(b);
}

function toNumber(v: unknown): number | undefined {
  if (typeof v === "number") return v;
  if (typeof v === "boolean") return v ? 1 : 0;
  if (typeof v === "string") {
    const n = Number(v);
    if (!Number.isNaN(n) && v.trim() !== "") return n;
  }
  return undefined;
}

function stringify(v: unknown): string {
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "boolean") return String(v);
  return JSON.stringify(v);
}
